atomix/cluster/src/test/java/io/atomix/raft/RaftRule.java

```diff
@@ -0,0 +1,521 @@
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+
+public final class RaftRule extends ExternalResource {
+
+  public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  private volatile int nextId;
+  private volatile List<RaftMember> members;
+  private Map<String, Long> memberLog;
+  private final List<RaftServer> servers = new ArrayList<>();
+  private volatile TestRaftProtocolFactory protocolFactory;
+  private volatile ThreadContext context;
+  private Path directory;
+  private final int nodeCount;
+  private volatile long highestCommit;
+  private final AtomicReference<CommitAwaiter> commitAwaiterRef = new AtomicReference<>();
+  private long position;
+
+  private RaftRule(final int nodeCount) {
+    this.nodeCount = nodeCount;
+  }
+
+  public static RaftRule withBootstrappedNodes(final int nodeCount) {
+    if (nodeCount < 1) {
+      throw new IllegalArgumentException("Expected to have at least one node to configure.");
+    }
+    return new RaftRule(nodeCount);
+  }
+
+  public static RaftRule withoutNodes() {
+    return new RaftRule(-1);
+  }
+
+  @Override
+  public Statement apply(final Statement base, final Description description) {
+    final var statement = super.apply(base, description);
+    return temporaryFolder.apply(statement, description);
+  }
+
+  @Override
+  protected void before() throws Throwable {
+    directory = temporaryFolder.newFolder().toPath();
+
+    position = 0;
+    members = new ArrayList<>();
+    memberLog = new ConcurrentHashMap<>();
+    nextId = 0;
+    context = new SingleThreadContext("raft-test-messaging-%d");
+    protocolFactory = new TestRaftProtocolFactory(context);
+
+    if (nodeCount > 0) {
+      createServers(nodeCount);
+    }
+  }
+
+  @Override
+  protected void after() {
+    servers.forEach(
+        s -> {
+          try {
+            if (s.isRunning()) {
+              s.shutdown().get(30, TimeUnit.SECONDS);
+            }
+          } catch (final Exception e) {
+            // its fine..
+          }
+        });
+    servers.clear();
+    context.close();
+    context = null;
+    members.clear();
+    nextId = 0;
+    protocolFactory = null;
+    highestCommit = 0;
+    commitAwaiterRef.set(null);
+    memberLog.clear();
+    memberLog = null;
+    position = 0;
+  }
+
+  /**
+   * Returns the next server address.
+   *
+   * @param type The startup member type.
+   * @return The next server address.
+   */
+  private RaftMember nextMember(final RaftMember.Type type) {
+    return new TestMember(nextNodeId(), type);
+  }
+
+  /**
+   * Returns the next unique member identifier.
+   *
+   * @return The next unique member identifier.
+   */
+  private MemberId nextNodeId() {
+    return MemberId.from(String.valueOf(++nextId));
+  }
+
+  /** Creates a set of Raft servers. */
+  private List<RaftServer> createServers(final int nodes) throws Exception {
```

Comment:
This method behaves strange, if members already has entries. Then new members would be added at the end, but the first entries would be used to create the servers.